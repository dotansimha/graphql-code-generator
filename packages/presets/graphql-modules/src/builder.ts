import {
  visit,
  DocumentNode,
  ObjectTypeDefinitionNode,
  ObjectTypeExtensionNode,
  Kind,
  TypeDefinitionNode,
  TypeExtensionNode,
  EnumTypeDefinitionNode,
  EnumTypeExtensionNode,
  InputObjectTypeDefinitionNode,
  InputObjectTypeExtensionNode,
  GraphQLSchema,
  isScalarType,
  InterfaceTypeDefinitionNode,
  InterfaceTypeExtensionNode,
} from 'graphql';
import { pascalCase } from 'change-case-all';
import {
  unique,
  withQuotes,
  buildBlock,
  pushUnique,
  concatByKey,
  uniqueByKey,
  createObject,
  collectUsedTypes,
  indent,
} from './utils';
import { ModulesConfig } from './config';
import { BaseVisitor } from '@graphql-codegen/visitor-plugin-common';

type RegistryKeys = 'objects' | 'inputs' | 'interfaces' | 'scalars' | 'unions' | 'enums';
type Registry = Record<RegistryKeys, string[]>;
const registryKeys: RegistryKeys[] = ['objects', 'inputs', 'interfaces', 'scalars', 'unions', 'enums'];
const resolverKeys: Array<Extract<RegistryKeys, 'objects' | 'enums' | 'scalars'>> = ['scalars', 'objects', 'enums'];

export function buildModule(
  name: string,
  doc: DocumentNode,
  {
    importNamespace,
    importPath,
    encapsulate,
    shouldDeclare,
    rootTypes,
    schema,
    baseVisitor,
  }: {
    importNamespace: string;
    importPath: string;
    encapsulate: ModulesConfig['encapsulateModuleTypes'];
    shouldDeclare: boolean;
    rootTypes: string[];
    baseVisitor: BaseVisitor;
    schema?: GraphQLSchema;
  }
): string {
  const picks: Record<RegistryKeys, Record<string, string[]>> = createObject(registryKeys, () => ({}));
  const defined: Registry = createObject(registryKeys, () => []);
  const extended: Registry = createObject(registryKeys, () => []);

  // List of types used in objects, fields, arguments etc
  const usedTypes = collectUsedTypes(doc);

  visit(doc, {
    ObjectTypeDefinition(node) {
      collectTypeDefinition(node);
    },
    ObjectTypeExtension(node) {
      collectTypeExtension(node);
    },
    InputObjectTypeDefinition(node) {
      collectTypeDefinition(node);
    },
    InputObjectTypeExtension(node) {
      collectTypeExtension(node);
    },
    InterfaceTypeDefinition(node) {
      collectTypeDefinition(node);
    },
    InterfaceTypeExtension(node) {
      collectTypeExtension(node);
    },
    ScalarTypeDefinition(node) {
      collectTypeDefinition(node);
    },
    UnionTypeDefinition(node) {
      collectTypeDefinition(node);
    },
    UnionTypeExtension(node) {
      collectTypeExtension(node);
    },
    EnumTypeDefinition(node) {
      collectTypeDefinition(node);
    },
    EnumTypeExtension(node) {
      collectTypeExtension(node);
    },
  });

  // Defined and Extended types
  const visited: Registry = createObject(registryKeys, key => concatByKey(defined, extended, key));

  // Types that are not defined or extended in a module, they come from other modules
  const external: Registry = createObject(registryKeys, key => uniqueByKey(extended, defined, key));

  //
  //
  //
  // Prints
  //
  //
  //

  // An actual output
  const imports = [`import * as ${importNamespace} from "${importPath}";`, `import * as gm from "graphql-modules";`];

  let content = [
    printDefinedFields(),
    printDefinedEnumValues(),
    printDefinedInputFields(),
    printSchemaTypes(usedTypes),
    printScalars(visited),
    printResolveSignaturesPerType(visited),
    printResolversType(visited),
    printResolveMiddlewareMap(),
  ]
    .filter(Boolean)
    .join('\n\n');

  if (encapsulate === 'namespace') {
    content =
      `${shouldDeclare ? 'declare' : 'export'} namespace ${baseVisitor.convertName(name, {
        suffix: 'Module',
        useTypesPrefix: false,
        useTypesSuffix: false,
      })} {\n` +
      (shouldDeclare ? `${indent(2)(imports.join('\n'))}\n` : '') +
      indent(2)(content) +
      '\n}';
  }

  return [...(!shouldDeclare ? imports : []), content].filter(Boolean).join('\n');

  /**
   * A dictionary of fields to pick from an object
   */
  function printDefinedFields() {
    return buildBlock({
      name: `interface DefinedFields`,
      lines: [...visited.objects, ...visited.interfaces].map(
        typeName =>
          `${typeName}: ${printPicks(typeName, {
            ...picks.objects,
            ...picks.interfaces,
          })};`
      ),
    });
  }

  /**
   * A dictionary of values to pick from an enum
   */
  function printDefinedEnumValues() {
    return buildBlock({
      name: `interface DefinedEnumValues`,
      lines: visited.enums.map(typeName => `${typeName}: ${printPicks(typeName, picks.enums)};`),
    });
  }

  function encapsulateTypeName(typeName: string): string {
    if (encapsulate === 'prefix') {
      return `${pascalCase(name)}_${typeName}`;
    }

    return typeName;
  }

  /**
   * A dictionary of fields to pick from an input
   */
  function printDefinedInputFields() {
    return buildBlock({
      name: `interface DefinedInputFields`,
      lines: visited.inputs.map(typeName => `${typeName}: ${printPicks(typeName, picks.inputs)};`),
    });
  }

  /**
   * Prints signatures of schema types with picks
   */
  function printSchemaTypes(types: string[]) {
    return types
      .filter(type => !visited.scalars.includes(type))
      .map(printExportType)
      .join('\n');
  }

  function printResolveSignaturesPerType(registry: Registry) {
    return [
      [...registry.objects, ...registry.interfaces]
        .map(name =>
          printResolverType(
            name,
            'DefinedFields',
            !rootTypes.includes(name) && defined.objects.includes(name) ? ` | '__isTypeOf'` : ''
          )
        )
        .join('\n'),
    ].join('\n');
  }

  function printScalars(registry: Registry) {
    if (!registry.scalars.length) {
      return '';
    }

    return [
      `export type ${encapsulateTypeName('Scalars')} = Pick<${importNamespace}.Scalars, ${registry.scalars
        .map(withQuotes)
        .join(' | ')}>;`,
      ...registry.scalars.map(scalar => {
        const convertedName = baseVisitor.convertName(scalar, {
          suffix: 'ScalarConfig',
        });
        return `export type ${encapsulateTypeName(convertedName)} = ${importNamespace}.${convertedName};`;
      }),
    ].join('\n');
  }

  /**
   * Aggregation of type resolver signatures
   */
  function printResolversType(registry: Registry) {
    const lines: string[] = [];

    for (const kind in registry) {
      const k = kind as RegistryKeys;
      if (registry.hasOwnProperty(k) && resolverKeys.includes(k as any)) {
        const types = registry[k];

        types.forEach(typeName => {
          if (k === 'enums') {
            return;
          } else if (k === 'scalars') {
            lines.push(`${typeName}?: ${encapsulateTypeName(importNamespace)}.Resolvers['${typeName}'];`);
          } else {
            lines.push(`${typeName}?: ${encapsulateTypeName(typeName)}Resolvers;`);
          }
        });
      }
    }

    return buildBlock({
      name: `export interface ${encapsulateTypeName('Resolvers')}`,
      lines,
    });
  }

  /**
   * Signature for a map of resolve middlewares
   */
  function printResolveMiddlewareMap() {
    const wildcardField = printResolveMiddlewareRecord(withQuotes('*'));
    const blocks: string[] = [buildBlock({ name: `${withQuotes('*')}?:`, lines: [wildcardField] })];

    // Type.Field
    for (const typeName in picks.objects) {
      if (picks.objects.hasOwnProperty(typeName)) {
        const fields = picks.objects[typeName];
        const lines = [wildcardField].concat(fields.map(field => printResolveMiddlewareRecord(field)));

        blocks.push(
          buildBlock({
            name: `${typeName}?:`,
            lines,
          })
        );
      }
    }

    return buildBlock({
      name: `export interface ${encapsulateTypeName('MiddlewareMap')}`,
      lines: blocks,
    });
  }

  function printResolveMiddlewareRecord(path: string): string {
    return `${path}?: gm.Middleware[];`;
  }

  function printResolverType(typeName: string, picksTypeName: string, extraKeys = '') {
    return `export type ${encapsulateTypeName(
      `${typeName}Resolvers`
    )} = Pick<${importNamespace}.${baseVisitor.convertName(typeName, {
      suffix: 'Resolvers',
    })}, ${picksTypeName}['${typeName}']${extraKeys}>;`;
  }

  function printPicks(typeName: string, records: Record<string, string[]>): string {
    return records[typeName].filter(unique).map(withQuotes).join(' | ');
  }

  function printTypeBody(typeName: string): string {
    const coreType = `${importNamespace}.${baseVisitor.convertName(typeName, {
      useTypesSuffix: true,
      useTypesPrefix: true,
    })}`;

    if (external.enums.includes(typeName) || external.objects.includes(typeName)) {
      if (schema && isScalarType(schema.getType(typeName))) {
        return `${importNamespace}.Scalars['${typeName}']`;
      }

      return coreType;
    }

    if (defined.enums.includes(typeName) && picks.enums[typeName]) {
      return `DefinedEnumValues['${typeName}']`;
    }

    if (defined.objects.includes(typeName) && picks.objects[typeName]) {
      return `Pick<${coreType}, DefinedFields['${typeName}']>`;
    }

    if (defined.interfaces.includes(typeName) && picks.interfaces[typeName]) {
      return `Pick<${coreType}, DefinedFields['${typeName}']>`;
    }

    if (defined.inputs.includes(typeName) && picks.inputs[typeName]) {
      return `Pick<${coreType}, DefinedInputFields['${typeName}']>`;
    }

    return coreType;
  }

  function printExportType(typeName: string): string {
    return `export type ${encapsulateTypeName(typeName)} = ${printTypeBody(typeName)};`;
  }

  //
  //
  //
  // Utils
  //
  //
  //

  function collectFields(
    node:
      | ObjectTypeDefinitionNode
      | ObjectTypeExtensionNode
      | InterfaceTypeDefinitionNode
      | InterfaceTypeExtensionNode
      | InputObjectTypeDefinitionNode
      | InputObjectTypeExtensionNode,
    picksObj: Record<string, string[]>
  ) {
    const name = node.name.value;

    if (node.fields) {
      if (!picksObj[name]) {
        picksObj[name] = [];
      }

      node.fields.forEach(field => {
        picksObj[name].push(field.name.value);
      });
    }
  }

  function collectValuesFromEnum(node: EnumTypeDefinitionNode | EnumTypeExtensionNode) {
    const name = node.name.value;

    if (node.values) {
      if (!picks.enums[name]) {
        picks.enums[name] = [];
      }

      node.values.forEach(field => {
        picks.enums[name].push(field.name.value);
      });
    }
  }

  function collectTypeDefinition(node: TypeDefinitionNode) {
    const name = node.name.value;

    switch (node.kind) {
      case Kind.OBJECT_TYPE_DEFINITION: {
        defined.objects.push(name);
        collectFields(node, picks.objects);
        break;
      }

      case Kind.ENUM_TYPE_DEFINITION: {
        defined.enums.push(name);
        collectValuesFromEnum(node);
        break;
      }

      case Kind.INPUT_OBJECT_TYPE_DEFINITION: {
        defined.inputs.push(name);
        collectFields(node, picks.inputs);
        break;
      }

      case Kind.SCALAR_TYPE_DEFINITION: {
        defined.scalars.push(name);
        break;
      }

      case Kind.INTERFACE_TYPE_DEFINITION: {
        defined.interfaces.push(name);
        collectFields(node, picks.interfaces);
        break;
      }

      case Kind.UNION_TYPE_DEFINITION: {
        defined.unions.push(name);
        break;
      }
    }
  }

  function collectTypeExtension(node: TypeExtensionNode) {
    const name = node.name.value;

    switch (node.kind) {
      case Kind.OBJECT_TYPE_EXTENSION: {
        collectFields(node, picks.objects);
        // Do not include root types as extensions
        // so we can use them in DefinedFields
        if (rootTypes.includes(name)) {
          pushUnique(defined.objects, name);
          return;
        }

        pushUnique(extended.objects, name);

        break;
      }

      case Kind.ENUM_TYPE_EXTENSION: {
        collectValuesFromEnum(node);
        pushUnique(extended.enums, name);
        break;
      }

      case Kind.INPUT_OBJECT_TYPE_EXTENSION: {
        collectFields(node, picks.inputs);
        pushUnique(extended.inputs, name);
        break;
      }

      case Kind.INTERFACE_TYPE_EXTENSION: {
        collectFields(node, picks.interfaces);
        pushUnique(extended.interfaces, name);
        break;
      }

      case Kind.UNION_TYPE_EXTENSION: {
        pushUnique(extended.unions, name);
        break;
      }
    }
  }
}
