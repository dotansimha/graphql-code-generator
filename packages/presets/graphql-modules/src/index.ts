import { Types } from '@graphql-codegen/plugin-helpers';
import { concatAST, isScalarType, parse } from 'graphql';
import { resolve, relative, join } from 'path';
import { groupSourcesByModule, stripFilename, normalize, isGraphQLPrimitive } from './utils';
import { buildModule } from './builder';
import { ModulesConfig } from './config';
import { BaseVisitor } from '@graphql-codegen/visitor-plugin-common';

export const preset: Types.OutputPreset<ModulesConfig> = {
  buildGeneratesSection: options => {
    const { baseOutputDir } = options;
    const { baseTypesPath, encapsulateModuleTypes } = options.presetConfig;

    const cwd = resolve(options.presetConfig.cwd || process.cwd());
    const importTypesNamespace = options.presetConfig.importTypesNamespace || 'Types';

    if (!baseTypesPath) {
      throw new Error(
        `Preset "graphql-modules" requires you to specify "baseTypesPath" configuration and point it to your base types file (generated by "typescript" plugin)!`
      );
    }

    if (!options.schemaAst || !options.schemaAst.extensions.sources) {
      throw new Error(`Preset "graphql-modules" requires to use GraphQL SDL`);
    }

    const sourcesByModuleMap = groupSourcesByModule(options.schemaAst!.extensions.sources, baseOutputDir);
    const modules = Object.keys(sourcesByModuleMap);

    const baseVisitor = new BaseVisitor(options.config, {});

    // One file with an output from all plugins
    const baseOutput: Types.GenerateOptions = {
      filename: resolve(cwd, baseOutputDir, baseTypesPath),
      schema: options.schema,
      documents: options.documents,
      plugins: [
        ...options.plugins,
        {
          'modules-exported-scalars': {},
        },
      ],
      pluginMap: {
        ...options.pluginMap,
        'modules-exported-scalars': {
          plugin: schema => {
            const typeMap = schema.getTypeMap();

            return Object.keys(typeMap)
              .map(t => {
                if (t && typeMap[t] && isScalarType(typeMap[t]) && !isGraphQLPrimitive(t)) {
                  const convertedName = baseVisitor.convertName(t);
                  return `export type ${convertedName} = Scalars["${t}"];`;
                }

                return null;
              })
              .filter(Boolean)
              .join('\n');
          },
        },
      },
      config: {
        ...(options.config || {}),
        enumsAsTypes: true,
      },
      schemaAst: options.schemaAst!,
    };

    const baseTypesFilename = baseTypesPath.replace(/\.(js|ts|d.ts)$/, '');
    const baseTypesDir = stripFilename(baseOutput.filename);

    // One file per each module
    const outputs: Types.GenerateOptions[] = modules.map(moduleName => {
      const filename = resolve(cwd, baseOutputDir, moduleName, options.presetConfig.filename);
      const dirpath = stripFilename(filename);
      const relativePath = relative(dirpath, baseTypesDir);
      const importPath = options.presetConfig.importBaseTypesFrom || normalize(join(relativePath, baseTypesFilename));
      const sources = sourcesByModuleMap[moduleName];

      const moduleDocument = concatAST(
        sources.map(source =>
          parse(source.body, {
            noLocation: true,
          })
        )
      );

      const shouldDeclare = filename.endsWith('.d.ts');

      return {
        filename,
        schema: options.schema,
        documents: [],
        plugins: [
          ...options.plugins.filter(p => typeof p === 'object' && !!p.add),
          {
            'graphql-modules-plugin': {},
          },
        ],
        pluginMap: {
          ...options.pluginMap,
          'graphql-modules-plugin': {
            plugin: schema =>
              buildModule(moduleName, moduleDocument, {
                importNamespace: importTypesNamespace,
                importPath,
                encapsulate: encapsulateModuleTypes || 'namespace',
                shouldDeclare,
                schema,
                baseVisitor,
                rootTypes: [
                  schema.getQueryType()?.name,
                  schema.getMutationType()?.name,
                  schema.getSubscriptionType()?.name,
                ].filter(Boolean),
              }),
          },
        },
        config: options.config,
        schemaAst: options.schemaAst,
      };
    });

    return [baseOutput].concat(outputs);
  },
};

export default preset;
