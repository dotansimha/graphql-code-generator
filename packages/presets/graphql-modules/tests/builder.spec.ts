import '@graphql-codegen/testing';
import { BaseVisitor } from '@graphql-codegen/visitor-plugin-common';
import { parse } from 'graphql';
import { buildModule } from '../src/builder';

const ROOT_TYPES = ['Query'];

const testDoc = parse(/* GraphQL */ `
  scalar DateTime
  scalar URL

  type Article {
    id: ID!
    title: String!
    text: String!
    author: User!
    comments: [Comment!]
    url: URL!
  }

  interface Node {
    id: ID!
  }

  union ArticleOrUser = Article | User

  input NewArticle {
    title: String!
    text: String!
  }

  extend type User {
    articles: [Article!]
  }

  extend type Query {
    articles: [Article!]
    articleById(id: ID!): Article
    articlesByUser(userId: ID!): [Article!]
  }

  enum UserKind {
    ADMIN
    WRITER
    REGULAR
  }
`);

const baseVisitor = new BaseVisitor({}, {});

test('should generate interface field resolvers', () => {
  const output = buildModule(
    'test',
    parse(/* GraphQL */ `
      interface BaseUser {
        id: ID!
        email: String!
      }

      type User implements BaseUser {
        id: ID!
        email: String!
      }

      type Query {
        me: BaseUser!
      }
    `),
    {
      importPath: '../types',
      importNamespace: 'core',
      encapsulate: 'none',
      shouldDeclare: false,
      rootTypes: ROOT_TYPES,
      baseVisitor,
    }
  );

  expect(output).toContain(`BaseUser: 'id' | 'email';`);
  expect(output).toContain(`export type BaseUser = Pick<core.BaseUser, DefinedFields['BaseUser']>;`);
  expect(output).toContain(`export type BaseUserResolvers = Pick<core.BaseUserResolvers, DefinedFields['BaseUser']>;`);
});

test('should generate interface extensions field resolvers ', () => {
  const output = buildModule(
    'test',
    parse(/* GraphQL */ `
      extend interface BaseUser {
        newField: String!
      }

      type Query {
        me: BaseUser!
      }
    `),
    {
      importPath: '../types',
      importNamespace: 'core',
      encapsulate: 'none',
      shouldDeclare: false,
      rootTypes: ROOT_TYPES,
      baseVisitor,
    }
  );

  expect(output).toContain(`BaseUser: 'newField';`);
  expect(output).toContain(`export type BaseUser = core.BaseUser`);
  expect(output).toContain(`export type BaseUserResolvers = Pick<core.BaseUserResolvers, DefinedFields['BaseUser']>;`);
});

test('should include import statement', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`
    import * as core from "../types";
  `);
});

test('should work with naming conventions', () => {
  const output = buildModule('test', parse(`type query_root { test: ID! } schema { query: query_root }`), {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toContain(`Pick<core.Query_RootResolvers, `);
  expect(output).toContain(`Pick<core.Query_Root,`);
});

test('encapsulate: should wrap correctly with namespace', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'namespace',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`export namespace TestModule {`);
  expect(output).toMatchSnapshot();
});

test('encapsulate: should wrap correctly with a declared namespace', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'namespace',
    shouldDeclare: true,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`declare namespace TestModule {`);
});

test('encapsulate: should wrap correctly with prefix', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'prefix',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toMatchSnapshot();
  expect(output).toContain(`export type Test_Article`);
  expect(output).toContain(`export type Test_User`);
  expect(output).toContain(`export type Test_Scalars`);
  expect(output).toContain(`export type Test_ArticleResolvers`);
  expect(output).toContain(`export interface Test_Resolvers`);
  expect(output).toContain(`export interface Test_MiddlewareMap`);
  expect(output).toContain(`interface DefinedFields {`);
  expect(output).toContain(`interface DefinedEnumValues {`);
  expect(output).toContain(`interface DefinedInputFields {`);
  expect(output).not.toBeSimilarStringTo(`export namespace Test {`);
});

test('should pick fields from defined and extended types', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`
    interface DefinedFields {
      Article: 'id' | 'title' | 'text' | 'author' | 'comments' | 'url';
      Query: 'articles' | 'articleById' | 'articlesByUser';
      User: 'articles';
      Node: 'id';
    };
  `);

  expect(output).toBeSimilarStringTo(`
    interface DefinedEnumValues {
      UserKind: 'ADMIN' | 'WRITER' | 'REGULAR';
    };
  `);

  expect(output).toBeSimilarStringTo(`
    interface DefinedInputFields {
      NewArticle: 'title' | 'text';
    };
  `);
});

test('should reexport used types but not defined in module', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`
    export type User = core.User;
  `);
  expect(output).toBeSimilarStringTo(`
    export type Comment = core.Comment;
  `);
});

test('should export partial types, only those defined in module or root types', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`
    export type Article = Pick<core.Article, DefinedFields['Article']>;
  `);
  expect(output).toBeSimilarStringTo(`
    export type Query = Pick<core.Query, DefinedFields['Query']>;
  `);
  expect(output).toBeSimilarStringTo(`
    export type UserKind = DefinedEnumValues['UserKind'];
  `);
  expect(output).toBeSimilarStringTo(`
    export type NewArticle = Pick<core.NewArticle, DefinedInputFields['NewArticle']>;
  `);
  expect(output).toBeSimilarStringTo(`
    export type Node = Pick<core.Node, DefinedFields['Node']>;
  `);
  expect(output).toBeSimilarStringTo(`
    export type ArticleOrUser = core.ArticleOrUser;
  `);
});

test('should export partial types of scalars, only those defined in module or root types', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`
    export type Scalars = Pick<core.Scalars, 'DateTime' | 'URL'>;
  `);

  // DateTime type should not be generated
  expect(output).not.toBeSimilarStringTo(`
    export type DateTime =
  `);
});

test('should use and export resolver signatures of types defined or extended in a module', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`
    export type ArticleResolvers = Pick<core.ArticleResolvers, DefinedFields['Article'] | '__isTypeOf'>;
  `);
  expect(output).toBeSimilarStringTo(`
    export type QueryResolvers = Pick<core.QueryResolvers, DefinedFields['Query']>;
  `);
  expect(output).toBeSimilarStringTo(`
    export type UserResolvers = Pick<core.UserResolvers, DefinedFields['User']>;
  `);
  expect(output).toBeSimilarStringTo(`
    export type DateTimeScalarConfig = core.DateTimeScalarConfig;
  `);
  expect(output).toBeSimilarStringTo(`
    export type UrlScalarConfig = core.UrlScalarConfig;
  `);
  // Interfaces should not have resolvers
  // We want Object types to have __isTypeOf
  expect(output).toBeSimilarStringTo(`
    export type NodeResolvers
  `);
  // Unions should not have resolvers
  // We want Object types to have __isTypeOf
  expect(output).not.toBeSimilarStringTo(`
    export type ArticleOrUserResolvers
  `);
});

test('should not generate resolver signatures of types that are not defined or extened by a module', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).not.toContain('CommentResolvers');
});

test('should generate an aggregation of individual resolver signatures', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toBeSimilarStringTo(`
    export interface Resolvers {
      Article?: ArticleResolvers;
      Query?: QueryResolvers;
      User?: UserResolvers;
      DateTime?: core.Resolvers['DateTime'];
      URL?: core.Resolvers['URL'];
    };
  `);
});

test('should generate a signature for ResolveMiddleware (with widlcards)', () => {
  const output = buildModule('test', testDoc, {
    importPath: '../types',
    importNamespace: 'core',
    encapsulate: 'none',
    shouldDeclare: false,
    rootTypes: ROOT_TYPES,
    baseVisitor,
  });

  expect(output).toContain(`import * as gm from "graphql-modules";`);

  expect(output).toBeSimilarStringTo(`
    export interface MiddlewareMap {
      '*'?: {
        '*'?: gm.Middleware[];
      };
      Article?: {
        '*'?: gm.Middleware[];
        id?: gm.Middleware[];
        title?: gm.Middleware[];
        text?: gm.Middleware[];
        author?: gm.Middleware[];
        comments?: gm.Middleware[];
        url?: gm.Middleware[];
      };
      User?: {
        '*'?: gm.Middleware[];
        articles?: gm.Middleware[];
      };
      Query?: {
        '*'?: gm.Middleware[];
        articles?: gm.Middleware[];
        articleById?: gm.Middleware[];
        articlesByUser?: gm.Middleware[];
      };
    };
  `);
});
