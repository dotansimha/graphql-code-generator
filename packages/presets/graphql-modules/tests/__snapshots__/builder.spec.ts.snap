// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encapsulate: should wrap correctly with namespace 1`] = `
"import * as core from \\"../types\\";
import * as gm from \\"graphql-modules\\";
export namespace TestModule {
  interface DefinedFields {
    Article: 'id' | 'title' | 'text' | 'author' | 'comments' | 'url';
    Query: 'articles' | 'articleById' | 'articlesByUser';
    User: 'articles';
    Node: 'id';
  };
  
  interface DefinedEnumValues {
    UserKind: 'ADMIN' | 'WRITER' | 'REGULAR';
  };
  
  interface DefinedInputFields {
    NewArticle: 'title' | 'text';
  };
  
  export type Article = Pick<core.Article, DefinedFields['Article']>;
  export type User = core.User;
  export type Comment = core.Comment;
  export type Node = Pick<core.Node, DefinedFields['Node']>;
  export type ArticleOrUser = core.ArticleOrUser;
  export type NewArticle = Pick<core.NewArticle, DefinedInputFields['NewArticle']>;
  export type Query = Pick<core.Query, DefinedFields['Query']>;
  export type UserKind = DefinedEnumValues['UserKind'];
  
  export type Scalars = Pick<core.Scalars, 'DateTime' | 'URL'>;
  export type DateTimeScalarConfig = core.DateTimeScalarConfig;
  export type UrlScalarConfig = core.UrlScalarConfig;
  
  export type ArticleResolvers = Pick<core.ArticleResolvers, DefinedFields['Article'] | '__isTypeOf'>;
  export type QueryResolvers = Pick<core.QueryResolvers, DefinedFields['Query']>;
  export type UserResolvers = Pick<core.UserResolvers, DefinedFields['User']>;
  export type NodeResolvers = Pick<core.NodeResolvers, DefinedFields['Node']>;
  
  export interface Resolvers {
    Article?: ArticleResolvers;
    Query?: QueryResolvers;
    User?: UserResolvers;
    DateTime?: core.Resolvers['DateTime'];
    URL?: core.Resolvers['URL'];
  };
  
  export interface MiddlewareMap {
    '*'?: {
      '*'?: gm.Middleware[];
    };
    Article?: {
      '*'?: gm.Middleware[];
      id?: gm.Middleware[];
      title?: gm.Middleware[];
      text?: gm.Middleware[];
      author?: gm.Middleware[];
      comments?: gm.Middleware[];
      url?: gm.Middleware[];
    };
    User?: {
      '*'?: gm.Middleware[];
      articles?: gm.Middleware[];
    };
    Query?: {
      '*'?: gm.Middleware[];
      articles?: gm.Middleware[];
      articleById?: gm.Middleware[];
      articlesByUser?: gm.Middleware[];
    };
  };
}"
`;

exports[`encapsulate: should wrap correctly with prefix 1`] = `
"import * as core from \\"../types\\";
import * as gm from \\"graphql-modules\\";
interface DefinedFields {
  Article: 'id' | 'title' | 'text' | 'author' | 'comments' | 'url';
  Query: 'articles' | 'articleById' | 'articlesByUser';
  User: 'articles';
  Node: 'id';
};

interface DefinedEnumValues {
  UserKind: 'ADMIN' | 'WRITER' | 'REGULAR';
};

interface DefinedInputFields {
  NewArticle: 'title' | 'text';
};

export type Test_Article = Pick<core.Article, DefinedFields['Article']>;
export type Test_User = core.User;
export type Test_Comment = core.Comment;
export type Test_Node = Pick<core.Node, DefinedFields['Node']>;
export type Test_ArticleOrUser = core.ArticleOrUser;
export type Test_NewArticle = Pick<core.NewArticle, DefinedInputFields['NewArticle']>;
export type Test_Query = Pick<core.Query, DefinedFields['Query']>;
export type Test_UserKind = DefinedEnumValues['UserKind'];

export type Test_Scalars = Pick<core.Scalars, 'DateTime' | 'URL'>;
export type Test_DateTimeScalarConfig = core.DateTimeScalarConfig;
export type Test_UrlScalarConfig = core.UrlScalarConfig;

export type Test_ArticleResolvers = Pick<core.ArticleResolvers, DefinedFields['Article'] | '__isTypeOf'>;
export type Test_QueryResolvers = Pick<core.QueryResolvers, DefinedFields['Query']>;
export type Test_UserResolvers = Pick<core.UserResolvers, DefinedFields['User']>;
export type Test_NodeResolvers = Pick<core.NodeResolvers, DefinedFields['Node']>;

export interface Test_Resolvers {
  Article?: Test_ArticleResolvers;
  Query?: Test_QueryResolvers;
  User?: Test_UserResolvers;
  DateTime?: Test_core.Resolvers['DateTime'];
  URL?: Test_core.Resolvers['URL'];
};

export interface Test_MiddlewareMap {
  '*'?: {
    '*'?: gm.Middleware[];
  };
  Article?: {
    '*'?: gm.Middleware[];
    id?: gm.Middleware[];
    title?: gm.Middleware[];
    text?: gm.Middleware[];
    author?: gm.Middleware[];
    comments?: gm.Middleware[];
    url?: gm.Middleware[];
  };
  User?: {
    '*'?: gm.Middleware[];
    articles?: gm.Middleware[];
  };
  Query?: {
    '*'?: gm.Middleware[];
    articles?: gm.Middleware[];
    articleById?: gm.Middleware[];
    articlesByUser?: gm.Middleware[];
  };
};"
`;
